File: src/main/java/math/cat/Categories.java
package math.cat;

import static java.util.Collections.singleton;
import static math.cat.Base.Map;
import static math.cat.Base.array;
import static math.cat.BasePair.Pair;
import static math.cat.BasePair.withLeft;
import static math.cat.Category.Category;
import static math.cat.Functions.id;
import static math.cat.Graph.Graph;
import static math.cat.Sets.Set;
import static math.cat.TypelessSetMorphism.compose;
import static math.cat.TypelessSetMorphism.forFunction;
import math.cat.Functions.Injection;
import math.cat.Sets.FactorSet;

import java.util.AbstractList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Some specific categories here.
 * 
 * @author Vlad Patryshev
 * All source code is stored on <a href="http://code.google.com/p/categories/">http://code.google.com/p/categories/</a>
 */
public final class Categories {
  private Categories() {}

  /**
   * Builds a category out of a segment of integers between 0 and n (not included).
   *
   * @param n number of elements
   * @return a new category
   */
  public static Category<Integer, Pair<Integer, Integer>> segment(int n) {
    return Category(PoSet.range(0, n - 1, 1));
  }

  /**
   * Builds a discrete category on a given set of objects. 
   * 
   * @param <T> object type
   * @param objects set of this category's objects
   * @return the category
   */
  @SuppressWarnings("unchecked")
  public static <T> Category<T, T> discreteCategory(Set<T> objects) {
    return Category(
        Graph(objects, id(objects), id(objects)),
        (Map<Pair<T, T>, T>) Collections.EMPTY_MAP);
  }

  /**
   * Empty category
   */
  public static final Category<Integer, Pair<Integer, Integer>> _0_ = segment(0);
  
  /**
   * Singleton category
   */
  public static final Category<Integer, Pair<Integer, Integer>> _1_ = segment(1);
  
  /**
   * Discrete 2-object category
   */
  public static final Category<String, String> _1plus1_ = discreteCategory(Set("a", "b"));
  
  /**
   * Category <b>2</b>: 2 objects linearly ordered
   */
  public static final Category<Integer, Pair<Integer, Integer>> _2_ = segment(2);
  
  /**
   * Category <b>3</b>: 3 objects linearly ordered
   */
  public static final Category<Integer, Pair<Integer, Integer>> _3_ = segment(3);
  
  /**
   * Category <b>4</b>: 4 objects linearly ordered
   */
  public static final Category<Integer, Pair<Integer, Integer>> _4_ = segment(4);
  
  /**
   * Category with 2 objects and 2 parallel arrows from one to another
   */
  @SuppressWarnings("unchecked")
  public static final Category<String, String> PARALLEL_PAIR =
      Category(Graph(Set("0", "1"), Map(array("a", "b"), array(Pair("0", "1"), Pair("0", "1")))), null);
  
  /**
   * Category <b>Z2</2> - a two-element monoid
   */
  public static final Category<String, String> Z2 = Category("(([1], {1: 1 -> 1, a: 1 -> 1}), {1 o 1 = 1, 1 o a = a, a o 1 = a, a o a = 1})");
  
  /**
   * "Split Monomorphism" category (see http://en.wikipedia.org/wiki/Morphism)
   * Two objects, and a split monomorphism from a to b 
   */
  public static final Category<String, String> SPLIT_MONO = 
      Category("(([a,b], {ab: a -> b, ba: b -> a, bb: b -> b}), {ab o ba = bb, ba o ab = a, bb o ab = ab, ba o bb = ba, bb o bb = bb})");
  
  /**
   * Commutative square category
   */
  public static final Category<String, String> SQUARE = Category("(([a,b,c,d], {ab: a -> b, ac: a -> c, bd: b -> d, cd: c -> d, ad: a -> d}), {bd o ab = ad, cd o ac = ad})");
  
  /**
   * Pullback category: a -> c <- b 
   */
  public static final Category<String, String> PULLBACK = Category("(([a,b,c], {ac: a -> c, bc: b -> c}), {})");
  
  /**
   * Pushout category: b <- a -> c 
   */
  public static final Category<String, String> PUSHOUT = Category("(([a,b,c], {ab: a -> b, ac: a -> c}), {})");
    
  /**
   * Sample W-shaped category: a -> b <- c -> d <- e
   */
  public static final Category<String, String> W = Category("(([a,b,c,d,e], {ab: a -> b, cb: c -> b, cd: c -> d, ed: e -> d}), {})");
  
  /**
   * Sample M-shaped category: a <- b -> c <- d -> e
   */
  public static final Category<String, String> M = Category("(([a,b,c,d,e], {ba: b -> a, bc: b -> c, dc: d -> c, de: d -> e}), {})");
  
  /**
   * Category of finite sets.
   */
  @SuppressWarnings("unchecked")
  public static Category<Set, TypelessSetMorphism> SETF =
      new Category<Set, TypelessSetMorphism>(BigSet.FINITE_SETS) {
        private final Set<TypelessSetMorphism> ALL_MORPHISMS =
            new BigSet<TypelessSetMorphism>() {
              public boolean contains(Object f) {
                return f instanceof SetMorphism;
              }
            };

        @Override
        public TypelessSetMorphism unit(Set x) {
          // I don't know wtf prevents dispatching TypelessSetMorphism.unit to TypelessSetMorphism...
          return TypelessSetMorphism.unitMorphism(x);
        }

        @Override
        public TypelessSetMorphism m(TypelessSetMorphism f, TypelessSetMorphism g) {
          assert f.codomain() == g.domain() : "Domain and codomain should match";
          return TypelessSetMorphism.compose(f, g);
        }

        @Override
        public Set<TypelessSetMorphism> arrows() {
          return ALL_MORPHISMS;
        }

        @Override
        protected void validate() { // it IS a category
        }

        @Override
        public String toString() {
          return "Category of all Java Sets";
        }

        @Override
        public Set<TypelessSetMorphism> arrows(Set x, Set y) {
          return TypelessSetMorphism.power(x, y);
        }

        @Override
        public boolean isIsomorphism(TypelessSetMorphism arrow) {
          return super.isIsomorphism(arrow);
        }

        @Override
        public boolean isMonomorphism(final TypelessSetMorphism arrow) {
          return new Predicate<Pair<Object, Object>>() {
            @Override
            public boolean eval(final Pair<Object, Object> p) {
              return equal(arrow.apply(p.x()), arrow.apply(p.y())) ? equal(p.x(), p.y()) : true;
            }
          }.forall(Sets.product(arrow.domain(), arrow.domain()));
        }

        @Override
        public boolean isEpimorphism(final TypelessSetMorphism arrow) {
          return new Predicate<Object>() {
            @Override
            public boolean eval(final Object y) {
              return new Predicate<Object>() {
                @Override
                public boolean eval(Object x) {
                  return y.equals(arrow.apply(x));
                }
              }.exists(arrow.domain());
            }
          }.forall(arrow.codomain());
        }

        @Override
        public Predicate<TypelessSetMorphism> factorsOnRight(TypelessSetMorphism g, TypelessSetMorphism f) {
          return super.factorsOnRight(g, f);
        }

        @Override
        public Predicate<TypelessSetMorphism> factorsOnLeft(TypelessSetMorphism g, TypelessSetMorphism f) {
          return super.factorsOnLeft(g, f);
        }

        @Override
        public Predicate<TypelessSetMorphism> factorsUniquelyOnLeft(TypelessSetMorphism f) {
          return super.factorsUniquelyOnLeft(f);
        }

        @Override
        public Predicate<TypelessSetMorphism> factorsUniquelyOnRight(TypelessSetMorphism f) {
          return super.factorsUniquelyOnRight(f);
        }

        @Override
        public TypelessSetMorphism equalizer(
            final TypelessSetMorphism f, final TypelessSetMorphism g) {
          assert f.domain() == g.domain() && f.codomain() == g.codomain();
          return TypelessSetMorphism.inclusion(f.domain(), new Predicate() {
            @Override
            public boolean eval(Object x) {
              return equal(f.apply(x), g.apply(x));
            }
          });
        }

        @Override
        public TypelessSetMorphism coequalizer(TypelessSetMorphism f, TypelessSetMorphism g) {
          assertParallelPair(f, g);
          FactorSet<Object> factorset = new FactorSet<Object>(f.codomain());
          for (Object x : f.domain()) {
            factorset.merge(f.apply(x), g.apply(x));
          }
          return TypelessSetMorphism.forFactorset(factorset);
        }

        @Override
        public TypelessSetMorphism coequalizer(final Iterable<TypelessSetMorphism> arrowsToEqualize, Set codomain) {
          if (!arrowsToEqualize.iterator().hasNext()) {
            return TypelessSetMorphism.unitMorphism(codomain);
          }
          FactorSet<Object> factorset = new FactorSet<Object>(codomain);
          Set domain = null;
          for (TypelessSetMorphism f : arrowsToEqualize) {
            assert equal(f.codomain(), codomain) : "Codomain should be " + codomain;
            if (domain == null) {
              domain = f.domain();
            } else {
              assert equal(f.domain(), domain) : "Arrows should have the same domain";
            }
          }
          TypelessSetMorphism f = null;
          for (TypelessSetMorphism g : arrowsToEqualize) {
            if (f != null) {
              for (Object x : g.domain()) {
                factorset.merge(f.apply(x), g.apply(x));
              }
            }
            f = g;
          }
          return TypelessSetMorphism.forFactorset(factorset);
        }

        @Override
        public Pair<TypelessSetMorphism, TypelessSetMorphism>
               product(Set x, Set y) {
          Set productSet = Sets.product(x, y);
          TypelessSetMorphism p = new TypelessSetMorphism(productSet, x) {
            @Override
            public Object apply(Object pair) {
              return ((Pair) pair).x();
            }
          };
          TypelessSetMorphism q = new TypelessSetMorphism(productSet, y) {
            @Override
            public Object apply(Object pair) {
              return ((Pair) pair).y();
            }
          };
          return Pair(p, q);
        }

        @Override
        public Pair<TypelessSetMorphism, TypelessSetMorphism> union(Set x, Set y) {
          Injection tagX = withLeft("x");
          Injection tagY = withLeft("y");
          Set taggedX = tagX.map(x);
          Set taggedY = tagY.map(y);
          Set unionSet = Sets.union(taggedX, taggedY);
          TypelessSetMorphism ix =
              compose(forFunction(x, taggedX, tagX), TypelessSetMorphism.inclusion(taggedX, unionSet));
          TypelessSetMorphism iy =
              compose(forFunction(y, taggedY, tagY), TypelessSetMorphism.inclusion(taggedY, unionSet));
          return Pair(ix, iy);
        }

        @Override
        public Pair<TypelessSetMorphism, TypelessSetMorphism> pullback(
            final TypelessSetMorphism f,
            final TypelessSetMorphism g) {
          Pair<TypelessSetMorphism, TypelessSetMorphism> product =
              product(f.domain(), g.domain());

          BinaryRelationship<Object, Object> pairIsGood =
              new BinaryRelationship<Object, Object>() {
                @Override
                public boolean eval(Pair<Object, Object> pair) {
                  return equal(f.apply(pair.x()), g.apply(pair.y()));
                }
              };

          TypelessSetMorphism pullbackToProduct =
              TypelessSetMorphism.inclusion(product.x().domain(), pairIsGood);
          return Pair(compose(pullbackToProduct, product.x()), compose(pullbackToProduct, product.y()));
        }

        @Override
        public Pair<TypelessSetMorphism, TypelessSetMorphism> pushout(
            TypelessSetMorphism f,
            TypelessSetMorphism g) {
          Pair<TypelessSetMorphism, TypelessSetMorphism> union =
              union(f.codomain(), g.codomain());
          TypelessSetMorphism coequalizer =
              coequalizer(compose(f, union.x()), compose(g, union.y()));
          return Pair(compose(union.x(), coequalizer), compose(union.y(), coequalizer));
        }

        @Override
        public Set initial() {
          return Collections.EMPTY_SET;
        }

        private final Set TERMINAL_SET = singleton(initial());

        @Override
        public Set terminal() {
          return TERMINAL_SET;
        }

        @Override
        public Pair<Set, List<TypelessSetMorphism>> degree(final Set x, final int n) {
          final Set<Map> allMaps = Sets.allMaps(Sets.numbers(n), x);
          return Pair(
              (Set) allMaps,
              (List<TypelessSetMorphism>) new AbstractList<TypelessSetMorphism>() {
                @Override
                public TypelessSetMorphism get(final int i) {
                  return new TypelessSetMorphism(allMaps, x) {
                    @Override
                    public Object apply(Object map) {
                      return ((Map) map).get(i);
                    }
                  };
                }

                @Override
                public int size() {
                  return n;
                }
              });
        }

        @Override
        public Category<Set, TypelessSetMorphism> op() {
          throw new UnsupportedOperationException("TODO(vpatryshev): try to implement it.");
        }

        @Override
        public Iterator<Set> iterator() {
          throw new UnsupportedOperationException("Not an enumerable entity.");
        }

        @Override
        public int size() {
          throw new UnsupportedOperationException("This is an infinite category.");
        }

        @Override
        public int hashCode() {
          return 7688721;
        }

        @Override
        public Set<Set> nodes() {
          return super.nodes();
        }

        @Override
        public Set d0(TypelessSetMorphism arrow) {
          return arrow.domain();
        }

        @Override
        public Set d1(TypelessSetMorphism arrow) {
          return arrow.codomain();
        }
      };
}

File: src/main/scala/math/cat/Categories.scala
package math.cat

import java.io.Reader
import java.util.Objects

import math.cat.Categories._
import math.sets.PoSet
import math.sets.Sets._
import scalakittens.Result._
import scalakittens.{Good, Result}

import scala.collection.{GenTraversableOnce, TraversableOnce, mutable}

private[cat] trait CategoryFactory {
  /**
    * Builds a category out of a segment of integers between 0 and n (not included).
    *
    * @param n number of elements
    * @return a new category
    */
  def segment(n: Int): Cat = {
    val numbers = fromPoset(s"_${n}_", PoSet.range(0, n, 1))
    val maybeSegment = convert2Cat(numbers)(
      _.toString,
      { case (a, b) ⇒ s"$a.$b" })
    maybeSegment.fold(identity, err ⇒ throw new InstantiationException(err.toString))
  }

  private def convert2Cat[O, A](source: Category)(
    object2string: source.Obj ⇒ String = (_: source.Obj).toString,
    arrow2string: source.Arrow ⇒ String = (_: source.Arrow).toString): Result[Cat] = {
    val stringToObject = source.objects map (o ⇒ object2string(o) → o) toMap
    val string2Arrow = source.arrows map (a ⇒ arrow2string(a) → a) toMap
    val objects = stringToObject.keySet
    val arrows = string2Arrow.keySet
    val d0 = (f: String) ⇒ object2string(source.d0(string2Arrow(f)))
    val d1 = (f: String) ⇒ object2string(source.d1(string2Arrow(f)))
    val ids = (o: String) ⇒ arrow2string(source.id(stringToObject(o)))
    val composition = (f: String, g: String) ⇒ source.m(string2Arrow(f), string2Arrow(g)) map arrow2string

    for {
      _ ← OKif(source.isFinite, "Need a finite category")
      _ ← OKif(objects.size == source.objects.size, "some objects have the same string repr")
      _ ← OKif(arrows.size == source.arrows.size, "some arrows have the same string repr")
      g ← Graph.build(source.name, objects, arrows, d0, d1)
      data: CategoryData = CategoryData(g)(
        ids.asInstanceOf[g.Node ⇒ g.Arrow],  // TODO: find a way to avoid casting
        composition.asInstanceOf[(g.Arrow, g.Arrow) ⇒ Option[g.Arrow]])

        c ← data.build
    } yield c.asInstanceOf[Cat]
  }

  /**
    * Builds a category out of a poset. Arrows are pairs (x,y) where x <= y.
    *
    * @tparam T poset element type
    * @param theName the name we give to the category we build
    * @param poset original poset
    * @return category based on he poset
    */
  def fromPoset[T](theName: String = "", poset: PoSet[T]): Category = {
    new Category {
      val graph = Graph.ofPoset(theName, poset)
      type Node = T
      type Arrow = (T, T)

      override def id(o: Obj): Arrow = arrow((o, o))

      override def m(f: Arrow, g: Arrow): Option[Arrow] = (f, g) match {
        case (f: (T, T), g: (T, T)) ⇒
          Option(f._1, g._2).filter(_ ⇒ f._2 == g._1) map arrow
      }
    }
  }

  def asCat(source: Category): Cat = convert2Cat(source)(_.toString, _.toString).getOrElse(
    throw new InstantiationException("Failed to convert to Cat")
  )

  /**
    * Builds a category given a limited (but sufficient) amount of data.
    * Objects have the same name as their identity arrows.
    *
    * @tparam T object and arrow type
    * @param objects           set of objects (same as identity arrows)
    * @param domain            maps arrows to domains
    * @param codomain          maps arrows to codomain
    * @param compositionSource source table of arrows composition (may be incomplete)
    * @return a newly-built category
    */
  def apply[T](
    name: String,
    objects: Set[T],
    domain: Map[T, T],
    codomain: Map[T, T],
    comp: Map[(T, T), T]): Result[Category] = {
    for {
      g ← Graph.build(name, objects, domain.keySet, domain, codomain)
      val partial = CategoryData.partial(g)(comp.asInstanceOf[Map[(g.Arrow, g.Arrow), g.Arrow]]) // same type actually
      c ← partial.build
    } yield c
  }

  /**
    * Builds a discrete category on a given set of objects.
    *
    * @tparam T object type
    * @param objects set of this category's objects
    * @return the category
    */
  def discrete[T](objects: Set[T]): Category = {
    new PartialData(Graph.discrete[T](objects, s"Discrete_${objects.size}")).build iHope
  }

  /**
    * Builds a category given a limited (but sufficient) amount of data.
    * Objects have the same name as their identities.
    *
    * @tparam T arrow and node type
    * @param name              category name
    * @param graph             the underlying graph
    * @param compositionSource source table of arrows composition (may be incomplete)
    * @return a newly-built category
    */

  private[cat] def addUnitsToGraph(graph: Graph): Graph = {

    val nodesOpt: Option[Set[Any]] = if (graph.isFinite) Some(graph.nodes.toSet) else None

    def isIdentity(f: Any): Boolean = nodesOpt map (_ contains f) getOrElse (graph contains f)

    new Graph {
      override val name = graph.name
      
      def nodes: Nodes = graph.nodes.asInstanceOf[Nodes]

      lazy val arrows: Arrows = (graph.nodes ++ graph.arrows).asInstanceOf[Arrows]

      def d0(f: Arrow): Node =
        if (isIdentity(f)) node(f) else node(graph.d0(graph.arrow(f)))

      def d1(f: Arrow): Node =
        if (isIdentity(f)) node(f) else node(graph.d1(graph.arrow(f)))
    }
  }

  def composablePairs(graph: Graph): Iterable[(graph.Arrow, graph.Arrow)] = {
    for (f ← graph.arrows; g ← graph.arrows if graph.follows(g, f)) yield (f, g)
  }

  /**
    * Factory method. Parses a string and builds a category from it.
    *
    * @param input input to parse
    * @return the category
    */
  def read(input: Reader): Result[Cat] = (new Parser).readCategory(input)

  /**
    * Factory method. Parses a string and builds a category from it.
    *
    * @param input the string to parse
    * @return the category
    */
  def read(input: CharSequence): Result[Cat] = (new Parser).readCategory(input)

  class Parser extends Graph.Parser {

    def readCategory(input: CharSequence): Result[Cat] = {
      val parseResult = parseAll(category, input)
      explain(parseResult)
    }

    def category: Parser[Result[Cat]] =
      (name ?) ~ "(" ~ graphData ~ (("," ~ multTable) ?) ~ ")" ^^ {
        case nameOpt ~ "(" ~ gOpt ~ mOpt ~ ")" ⇒ {
          val name = nameOpt.getOrElse("c")
          val graphOpt = gOpt.map(_.build(name))

          mOpt match {
            case None ⇒
              buildCategory(graphOpt, Map.empty)
            case Some("," ~ m) ⇒
              buildCategory(graphOpt, m)
            case Some(garbage) ⇒ Result.error(s"bad data: $garbage")
          }
        }
      }

    private def buildCategory(
      gOpt: Result[Graph],
      multTable: Map[(String, String), String]): Result[Cat] = {
      for {
        g: Graph ← gOpt
        raw ← new PartialData(g) {
//          override val graph = g
          override val compositionSource = multTable.asInstanceOf[Map[(graph.Arrow, graph.Arrow), graph.Arrow]]
        }.build
        cat ← convert2Cat(raw)()
      } yield cat
    }

    def multTable: Parser[Map[(String, String), String]] = "{" ~ repsep(multiplication, ",") ~ "}" ^^ { case "{" ~ m
      ~ "}" ⇒ Map() ++ m
    }

    def multiplication: Parser[((String, String), String)] = {
      word ~ ("o"|"∘") ~ word ~ "=" ~ word ^^ { case g ~ o ~ f ~ "=" ~ h ⇒ ((f, g), h)
      }
    }

    def readCategory(input: Reader): Result[Cat] = {
      val parseResult = parseAll(category, input)
      explain(parseResult)
    }
  }

}


/**
  * A bunch of specific categories in this object
  */
object Categories extends CategoryFactory {

  type Cat = Category {
    type Node = String
    type Arrow = String
  }

  /**
    * Empty category
    */
  lazy val _0_ : Cat = segment(0)

  /**
    * Singleton category
    */
  lazy val _1_ : Cat = segment(1)

  /**
    * Discrete 2-object category
    */
  lazy val _1plus1_ : Cat = asCat(discrete(Set("a", "b")))

  /**
    * Category <b>2</b>: 2 objects linearly ordered
    */
  lazy val _2_ : Cat = segment(2)

  /**
    * Category <b>3</b>: 3 objects linearly ordered
    */
  lazy val _3_ : Cat = segment(3)

  /**
    * Category <b>4</b>: 4 objects linearly ordered
    */
  lazy val _4_ : Cat = segment(4)

  /**
    * Category <b>5</b>: 5 objects linearly ordered
    */
  lazy val _5_ : Cat = segment(5)

  /**
    * Category with 2 objects and 2 parallel arrows from one to another
    */
  lazy val ParallelPair = category"ParallelPair:({0, 1}, {a:0→1, b:0→1})"

  /**
    * Category <b>Z2</2> - a two-element monoid
    */
  lazy val Z2 = category"Z2: ({1}, {1: 1 → 1, a: 1 → 1}, {1 ∘ 1 = 1, 1 ∘ a = a, a ∘ 1 = a, a ∘ a = 1})"

  lazy val Z3 = category"Z3: ({0}, {0: 0 → 0, 1: 0 → 0, 2: 0 → 0}, {1 ∘ 1 = 2, 1 ∘ 2 = 0, 2 ∘ 1 = 0, 2 ∘ 2 = 1})"

  /**
    * "Split Monomorphism" category (see http://en.wikipedia.org/wiki/Morphism)
    * Two objects, and a split monomorphism from a to b
    */
  lazy val SplitMono =
    category"SplitMono: ({a,b}, {ab: a → b, ba: b → a, bb: b → b}, {ab ∘ ba = bb, bb ∘ bb = bb})"

  /**
    * Commutative square category
    */
  lazy val Square = category"Square:({a,b,c,d}, {ab: a → b, ac: a → c, bd: b → d, cd: c → d, ad: a → d})"

  /**
    * Pullback category: a → c ← b
    */
  lazy val Pullback = category"Pullback:({a,b,c}, {ac: a → c, bc: b → c})"

  /**
    * Pushout category: b ← a → c
    */
  lazy val Pushout = category"Pushout:({a,b,c}, {ab: a → b, ac: a → c})"

  /**
    * Pushout3 category: b ← a → c
    */
  lazy val Pushout4 = category"Pushout4:({a,b,c,d,e}, {ab: a → b, ac: a → c, ad: a → d, ae: a → e})"

  /**
    * Sample W-shaped category: a → b ← c → d ← e
    */
  lazy val W = category"W:({a,b,c,d,e}, {ab: a → b, cb: c → b, cd: c → d, ed: e → d})"

  /**
    * Sample M-shaped category: a ← b → c ← d → e
    */
  lazy val M = category"M:({a,b,c,d,e}, {ba: b → a, bc: b → c, dc: d → c, de: d → e})"


  /**
    * A segment of simplicial category.
    * Represents three sets (empty, singleton and two-point) and
    * all their possible functions.
    */
  lazy val HalfSimplicial: Cat = asCat(apply("HalfSimplicial",
    Set("0", "1", "2"),
    Map("0_1" → "0", "0_2" → "0", "2_1" → "2", "2_a" → "2", "2_b" → "2", "a" → "1", "b" → "1", "2_swap" →
      "2"), // d0
    Map("0_1" → "1", "0_2" → "2", "2_1" → "1", "2_a" → "2", "2_b" → "2", "a" → "2", "b" → "2", "2_swap" →
      "2"), // d1
    Map(("0_1", "a") → "0_2",
      ("0_1", "b") → "0_2",
      ("2_1", "a") → "2_a",
      ("2_1", "b") → "2_b",
      ("a", "2_swap") → "b",
      ("a", "2_a") → "a",
      ("b", "2_swap") → "a",
      ("b", "2_a") → "a",
      ("b", "2_b") → "b",
      ("2_swap", "2_swap") → "2",
      ("2_swap", "2_a") → "2_a",
      ("2_swap", "2_b") → "2_b",
      ("2_a", "2_a") → "2_a",
      ("2_b", "2_b") → "2_b",
      ("2_a", "2_swap") → "2_b",
      ("2_b", "2_swap") → "2_a"
    )
  ).
    getOrElse(throw new InstantiationException("Bad semisimplicial?")))
  
  lazy val AAAAAA = category"AAAAAA: ({1,2,3,4,5,6}, {12: 1 → 2, 23: 2 → 3, 34: 3 → 4, 45: 4 → 5, 56: 5 → 6, 61: 6 → 1})"

  lazy val NaturalNumbers: Category = fromPoset("ℕ", PoSet.ofNaturalNumbers)

  lazy val SomeKnownCategories = List(
    _0_, _1_, _3_, ParallelPair, SplitMono, W, M, Z3)

  lazy val KnownCategories: List[Category] = List(
    _0_, _1_, _2_, _3_, _4_, _5_, _1plus1_,
    ParallelPair, Pullback, Pushout, Pushout4, SplitMono, Square,
    M, W,
    Z2, Z3,
//    AAAAAA,
    HalfSimplicial, NaturalNumbers).sortBy(_.arrows.size)

  lazy val KnownFiniteCategories: List[Category] =
    KnownCategories filter (_.isFinite)

  implicit class CategoryString(val sc: StringContext) extends AnyVal {
    def category(args: Any*): Cat = {
      val strings = sc.parts.iterator
      val expressions = args.iterator
      var buf = new StringBuffer(strings.next)
      while (strings.hasNext) {
        buf append expressions.next
        buf append strings.next
      }
      read(buf) match {
        case Good(c) ⇒ c
        case bad ⇒ throw new InstantiationException(bad.errorDetails.mkString)
      }
    }
  }
}
